Nintendo IdP addition.

robfusion opened the pull request.

robotdan commented on site/docs/v1/tech/identity-providers/nintendo.adoc:
```diff
@@ -0,0 +1,63 @@
+
+Once you have completed this configuration you will be able to enable the Nintendo login button for one or more FusionAuth Applications. Below is an example login page with the Nintendo Identity Provider enabled.
+
+image::identity-providers/login-nintendo.png[Nintendo Login]
```
I think the rest of the IdP examples have a cropped image here.

robfusion replied:
I did actually crop it, but accidentally saved the cropped version in _site instead of site. Fixing it now.

robfusion replied:
Should be fixed now.

robotdan commented on site/docs/v1/tech/identity-providers/nintendo.adoc:
```diff
@@ -0,0 +1,63 @@
+
+== Create a Nintendo Developer Account
+
+Follow instructions for obtaining a `Client id` and `Client secret` from the official Nintendo documentation.
```
Follow the naming of the field names, this should be `Client Id`.

robotdan replied:
Since the heading says "create a Nintendo Developer Account" it seems we should have a bit more here even if it is a sentence that says 

> Log into XYZ and create a developer account and enable the developer account for Login with Nintendo, and collect the Client Id and Client secret. Contact Nintendo Developer support for additional information. 

Or something simple like that.

robfusion replied:
I took a shot. There is not much info available though.

robotdan commented on site/docs/v1/tech/lambdas/nintendo-reconcile.adoc:
```diff
@@ -42,7 +37,7 @@ Navigate to [breadcrumb]#Settings -> Identity Providers# and select your existin
 
 == Example lambda
 
-There is not a example Nintendo reconcile lambda provided by FusionAuth. The following is an example lambda you can use as a starting point.
+There is no example Nintendo reconcile lambda provided by FusionAuth. The following is an example lambda you can use as a starting point.
```
We have other IdPs that don't ship with a default lambda, such as Epic and Steam. We should either follow those examples, or you could also add this type of example to other IdPs that do not ship a default reconcile lambda.

robfusion replied:
Yeah, that was already there so I just left it. I'll remove it to match the others.

robfusion replied:
Removed.

robotdan reviewed (commented):
Are there changes to me made int he IdP APIs, or the IdP Login API?

robfusion commented:
> 

Yeah, I totally missed the IdP APIs working on those now and I forget that we added an Options tab like in OIDC that needs to be documented as well. Working on those.

robotdan commented on site/docs/src/json/identity-providers/nintendo-response.json:
```diff
@@ -0,0 +1,33 @@
+    "debug" : false,
+    "emailClaim" : "email",
+    "enabled" : true,
+    "id" : "7764b5c7-165b-4e7e-94aa-02ebe2a0a5fb",
```
This is not the correct Id, it should be `b0ac2e16-d4af-483e-98c8-7f6693610665`.

robfusion replied:
I just thought it was a dummy value as an example. Didn't realize it needed to be something specific. I'll change it.

robotdan commented on site/docs/src/json/identity-providers/sonypsn-request.json:
```diff
@@ -6,7 +6,7 @@
       }
     },
-    "buttonText" : "Login with Epic Games",
+    "buttonText" : "Login with Sony PlayStation Network",
```
Not a big deal, but the default value is `Login with Playstation Network`.

robfusion replied:
Yeah, just saw it was Epic in the sony example and figured I would fix it while I was there. I'll change it.

robotdan commented on site/docs/v1/tech/apis/identity-providers/_oauth-idp-request-body.adoc:
```diff
@@ -67,6 +72,14 @@ ifeval::["{idp_type}" == "Steam"]
 
+ifeval::["{idp_type}" == "Nintendo"]
+[field]#identityProvider.uniqueIdClaim# [type]#[String]# [optional]#Optional# [default]#defaults to `id`#::
+An optional configuration to modify the expected name of the claim returned by the {idp_display_name} Identity Provider that contains the user id.
```
`user id` -> `user Id`.

robfusion replied:
should be fixed

robotdan commented on site/docs/v1/tech/apis/identity-providers/_oauth-idp-response-body.adoc:
```diff
@@ -77,6 +82,14 @@ ifeval::["{idp_type}" == "Steam"]
 
+ifeval::["{idp_type}" == "Nintendo"]
+[field]#identityProvider.uniqueIdClaim# [type]#[String]#::
+The optional configuration to modify the expected name of the claim returned by the {idp_display_name} Identity Provider that contains the user id.
```
`user id` -> `user Id`

robfusion replied:
should be fixed

robotdan commented on site/docs/v1/tech/apis/identity-providers/_oauth-idp-response-body.adoc:
```diff
@@ -77,6 +82,14 @@ ifeval::["{idp_type}" == "Steam"]
 endif::[]
 
+ifeval::["{idp_type}" == "Nintendo"]
+[field]#identityProvider.uniqueIdClaim# [type]#[String]#::
```
Looks like we are missing the `uniqueIdClaim` and `usernameClaim` claim for OpenID Connect, can you add them?

robfusion replied:
yep

robfusion replied:
should be fixed

robotdan commented on site/docs/v1/tech/apis/identity-providers/nintendo.adoc:
```diff
@@ -0,0 +1,46 @@
+---
+layout: doc
+title: Nintendo Identity Provider APIs
+description: APIs for creating, retrieving, updating and deleting Nintendo identity providers
```
We should use similar language to that of Steam, Twitch, or other typed IdPs. Specifically, you cannot have more than one Nintendo IdP, so we should refer to this in a singular fashion.

robfusion replied:
should be fixed

robotdan commented on site/docs/v1/tech/apis/identity-providers/nintendo.adoc:
```diff
@@ -0,0 +1,46 @@
+
+The Nintendo identity provider type will use the Nintendo OAuth login API. It will also provide a `Login with Nintendo` button on FusionAuth's login page.
+
+FusionAuth will leverage the `/users/me` API that is part of the Nintendo specification. The email address and username returned in the response will be used to create or lookup the existing User. Additional claims from the response can be used to reconcile the User in FusionAuth by using an Nintendo Reconcile Lambda. Unless you assign a reconcile lambda to this provider or configure the IdP opti ...
```
> The email address and username returned in the response will be used to create or lookup the existing User.

This is true if it is returned, and based upon the linking strategy.

robfusion replied:
should be fixed

robotdan commented on site/docs/v1/tech/apis/identity-providers/nintendo.adoc:
```diff
@@ -0,0 +1,46 @@
+:hide_token_map_deprecation!:
+
+:idp_display_name: Nintendo
+:idp_id: 7764b5c7-165b-4e7e-94aa-02ebe2a0a5fb
```
This is the wrong Id, it should be `b0ac2e16-d4af-483e-98c8-7f6693610665`.

robfusion replied:
should be fixed

robotdan commented on site/docs/v1/tech/identity-providers/sony.adoc:
```diff
@@ -35,7 +35,7 @@ link:https://partners.playstation.net/[Sony Playstation Network - Getting Starte
 == Create a Sony Playstation Network Identity Provider
 
-:identity_providerPlaystation:
+:identity_provider: Playstation
```
Not sure where this is used, but should it be `Sony PlayStation Network` that is what our display name is in FusionAuth.

robfusion replied:
Yeah, I just noticed it was messed up while I was looking it over as an example. I'll change it to Sony Playstation Network

mooreds commented:
missing images (maybe handled in 1380?)

```
[2022-05-11 15:26:20] ERROR `/assets/img/docs/identity-providers/identity-provider-nintendo-add-options.png' not found.
[2022-05-11 15:26:20] ERROR `/assets/img/docs/identity-providers/identity-provider-nintendo-add.png' not found.
[2022-05-11 15:26:20] ERROR `/assets/img/docs/identity-providers/identity-provider-nintendo-add.png' not found.
[2022-05-11 15:26:20] ERROR `/assets/img/docs/identity-providers/identity-provider-nintendo-add-options.png' not found.
```

* icon in nav is a 4
* icon in nav under api is missing (could just be firefox, if it works for you, ignore this and I can check it out when it is published)

identity provider API notes

* button text is marked as required in API but not in the UX (maybe expected bc we provide a default?)

* since this was introduced in 1.36, doesn't make sense to have identityProvider.tenantConfiguration be 'since 1.32'. There are others fields in that same boat, both in request and response. pendingIdPLinkId is 'since 1.28'
* looks like 7764b5c7-165b-4e7e-94aa-02ebe2a0a5fb is the PSN idp (want to change that in the json as well as the doco)

identity provider notes (not API)

* developer portal link has to slashes at the end: https://developer.nintendo.com//
* image under 'Create a Nintendo Identity Provider' is missing.
* looks like 'scope_addon' is not defined, {scope_addon} is present in the scope field defn
* image under 'Options' is missing.
* Other options are missing from the identity provider page (not the API, but the top level)

robfusion commented:
> 

Fixed missing screenshots. I forgot to do a git add so they looked fine in my local UI. :)

I don't know what "icon in nav is a 4" means

Hmm, icon is missing for me now too, but I have no idea why. 

Button text required in API but not UI. Seems to be the same for all. I assume it's because we pre-populate in the UI.

I wasn't sure about the versioning up specific fields. I can remove all the versioning, but it seemed harmless and would require repeating a bunch of field text or making complicated logic to exclude the versioning just for Nintendo.

I think Daniel mentioned the IdP id thing too. I changed it everywhere he mentioned it. I thought it was just a placeholder so didn't really matter.

Fixed missing screenshots I believe.

Not sure which missing Options you are referring to. I see three options unique id claim, email claim, and username claim. That should be all for Nintendo.

scope_addon is only used for Twitch

mooreds commented:
The icon was missing in the API; was a typo. I fixed and pushed ( https://github.com/FusionAuth/fusionauth-site/pull/1379/commits/b75b4780be8e534e19b9c659590a0bbac5fe7d4c ).
However, I'm still missing the correct icon. I don't think we pay for 'pro' font awesome icons: https://fontawesome.com/icons/game-console-handheld?s=solid Consider https://fontawesome.com/icons/gamepad?s=solid instead, maybe?

re: 1.32 stuff. I prefer it to not be present as it seems weird. See site/docs/v1/tech/apis/identity-providers/_identity-provider-reconcileid-request-parameter.adoc for an example of how to do this.

re: icon in nav is a 4. Here's a screenshot of my nav (in FF and in Chrome).
<img width="322" alt="Screen Shot 2022-05-12 at 10 31 14 AM" src="https://user-images.githubusercontent.com/91825/168125534-7af90622-0381-482d-a24d-3bbcc2440956.png">
<img width="327" alt="Screen Shot 2022-05-12 at 10 31 37 AM" src="https://user-images.githubusercontent.com/91825/168125537-c9916c78-05f9-4f07-9bd0-133a0e01daf5.png">


`Nintendo home` image is still missing:

<img width="686" alt="Screen Shot 2022-05-12 at 10 37 02 AM" src="https://user-images.githubusercontent.com/91825/168125543-8e3d255c-ca4e-49c9-8128-b9d7d00b1d71.png">


See my screenshot for the scope options issue (look for `scope_addon`; I think you need to define it as an empty string).
 
<img width="831" alt="Screen Shot 2022-05-12 at 10 34 49 AM" src="https://user-images.githubusercontent.com/91825/168125540-6a8078e6-7a78-47e2-8a91-d3db4795980b.png">

robfusion commented:
> The icon was missing in the API; was a typo. I fixed and pushed ( [b75b478](https://github.com/FusionAuth/fusionauth-site/commit/b75b4780be8e534e19b9c659590a0bbac5fe7d4c) ). However, I'm still missing the correct icon. I don't think we pay for 'pro' font awesome icons: https://fontawesome.com/icons/game-console-handheld?s=solid Consider https://fontawesome.com/icons/gamepad?s=solid instead, maybe?
> 
> re: 1.32 stuff. I prefer it to not be present as it seems weird. See site/docs/v1/tech/apis/identity-providers/_identity-provider-reconcileid-request-parameter.adoc for an example of how to do this.
> 
> re: icon in nav is a 4. Here's a screenshot of my nav (in FF and in Chrome). <img alt="Screen Shot 2022-05-12 at 10 31 14 AM" width="322" src="https://user-images.githubusercontent.com/91825/168125534-7af90622-0381-482d-a24d-3bbcc2440956.png"> <img alt="Screen Shot 2022-05-12 at 10 31 37 AM" width="327" src="https://user-images.githubusercontent.com/91825/168125537-c9916c78-05f9-4f07-9bd0-133a0e01daf5.png">
> 
> `Nintendo home` image is still missing:
> 
> <img alt="Screen Shot 2022-05-12 at 10 37 02 AM" width="686" src="https://user-images.githubusercontent.com/91825/168125543-8e3d255c-ca4e-49c9-8128-b9d7d00b1d71.png">
> 
> See my screenshot for the scope options issue (look for `scope_addon`; I think you need to define it as an empty string).
> 
> <img alt="Screen Shot 2022-05-12 at 10 34 49 AM" width="831" src="https://user-images.githubusercontent.com/91825/168125540-6a8078e6-7a78-47e2-8a91-d3db4795980b.png">

Ok, I think I got all of these. I'm still not sure if the icon is rendering right. It seems to work sometimes and not other times for some reason.
